File: android/control-proxy/src/main/kotlin/dev/jasonpearson/automobile/ctrlproxy/models/ViewHierarchy.kt
package dev.jasonpearson.automobile.ctrlproxy.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Complete view hierarchy representation matching the structure expected by AutoMobile test
 * framework
 */
@Serializable
data class ViewHierarchy
@JvmOverloads
constructor(
  val updatedAt: Long = System.currentTimeMillis(),
  val packageName: String? = null,
  /** Android user that owns the accessibility service that captured this hierarchy. */
  val userId: Int? = null,
  val hierarchy: UIElementInfo? = null,
  val windowInfo: WindowInfo? = null,
  val windows: List<WindowInfo>? = null,
  val contentHiddenRegions: List<ContentHiddenRegion>? = null,
  val intentChooserDetected: Boolean? = null,
  val notificationPermissionDetected: Boolean? = null,
  @SerialName("accessibility-focused-element")
  val accessibilityFocusedElement: UIElementInfo? = null, // Element with TalkBack cursor
  val ctrlProxyIncomplete: Boolean? = null,
  val error: String? = null, // For error cases like locked screen
  val screenWidth: Int? = null,
  val screenHeight: Int? = null,
  val rotation: Int? = null, // 0=portrait, 1=landscape90, 2=reverse, 3=landscape270
  val systemInsets: SystemInsetsInfo? = null,
  val insets: ObservationInsetsInfo? = null,
  val wakefulness: String? = null, // "Awake", "Asleep", or "Dozing"
  val foregroundActivity: String? = null, // e.g. "com.example.app/.MainActivity"
  val density: Int? = null, // Display density in DPI
  val sdkInt: Int? = null, // Android API level (e.g. 34)
  val deviceModel: String? = null, // e.g. "Pixel 8"
  val isEmulator: Boolean? = null, // Whether running on an emulator
  // Additive scale metadata (#4548): the ratio between the bounds units reported in this
  // hierarchy and physical screenshot pixels. Android accessibility bounds and screenshots
  // are BOTH physical pixels, so the truthful value is exactly 1 — reported explicitly so
  // both platforms carry the same metadata shape (iOS reports UIScreen.nativeScale).
  val nativeScale: Float? = null,
  val pixelWidth: Int? = null, // Physical screenshot pixel width (== screenWidth on Android)
  val pixelHeight: Int? = null, // Physical screenshot pixel height (== screenHeight on Android)
  /** Structured reasons why this snapshot is partial or unavailable. */
  val truncationReasons: List<String>? = null,
)

@Serializable
data class ContentHiddenRegion(
  val bounds: ElementBounds,
  val reason: String,
  val areaPercent: Int,
)

File: ios/control-proxy/Sources/CtrlProxyRewrite/Models/ViewHierarchy.swift
import Foundation

/// View hierarchy structure (matching Android's ViewHierarchy). The `hierarchy_update`
/// / `request_hierarchy` result payload; the deep element tree the TS client renders.
public struct ViewHierarchy: Codable, Sendable {
    public let updatedAt: Int64
    public let packageName: String?
    public let hierarchy: UIElementInfo?
    public let windowInfo: WindowInfo?
    public let windows: [WindowInfo]?
    public let screenScale: Float?
    public let screenWidth: Int?
    public let screenHeight: Int?
    /// Ratio between the point-space bounds reported here and the physical screenshot
    /// pixels (`UIScreen.nativeScale`). Distinct from `screenScale` (`UIScreen.scale`):
    /// Display Zoom changes `nativeScale` but not `scale`, and screenshots are rendered
    /// at native scale, so this converts reported bounds to screenshot pixels (#4548).
    public let nativeScale: Double?
    /// Physical screenshot pixel width: `round(screenWidth * nativeScale)` (#4548).
    public let pixelWidth: Int?
    /// Physical screenshot pixel height: `round(screenHeight * nativeScale)` (#4548).
    public let pixelHeight: Int?
    /// Device display rotation captured with this hierarchy: Android-compatible 0...3.
    public let rotation: Int?
    public let systemInsets: EdgeInsetsInfo?
    public let insets: ObservationInsetsInfo
    public let error: String?
    public let fallbackToSpringboard: Bool?

    public init(
        updatedAt: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        packageName: String? = nil,
        hierarchy: UIElementInfo? = nil,
        windowInfo: WindowInfo? = nil,
        windows: [WindowInfo]? = nil,
        screenScale: Float? = nil,
        screenWidth: Int? = nil,
        screenHeight: Int? = nil,
        nativeScale: Double? = nil,
        pixelWidth: Int? = nil,
        pixelHeight: Int? = nil,
        rotation: Int? = nil,
        systemInsets: EdgeInsetsInfo? = nil,
        insets: ObservationInsetsInfo = .unavailable,
        error: String? = nil,
        fallbackToSpringboard: Bool? = nil
    ) {
        self.updatedAt = updatedAt
        self.packageName = packageName
        self.hierarchy = hierarchy
        self.windowInfo = windowInfo
        self.windows = windows
        self.screenScale = screenScale
        self.screenWidth = screenWidth
        self.screenHeight = screenHeight
        self.nativeScale = nativeScale
        self.pixelWidth = pixelWidth
        self.pixelHeight = pixelHeight
        self.rotation = rotation
        self.systemInsets = systemInsets
        self.insets = insets
        self.error = error
        self.fallbackToSpringboard = fallbackToSpringboard
    }
}
